packages/analytics-js-integrations/src/integrations/BingAds/utils.js

```diff
@@ -91,10 +95,103 @@ const buildEcommPayload = message => {
   return { ...ecommPayload, ...payload };
 };
 
+/**
+ * Formats the email and hashes the email
+ * Docs for formatting email
+ * https://help.ads.microsoft.com/apex/index/3/en/60178#:~:text=do%20so%20manually.-,Format%20and%20hash%20the%20data,-Format%20the%20data
+ * @param {*} emailString
+ * @returns  hash of finalised email and if it is invalid returns undefined
+ */
+const formatAndHashEmailAddress = emailString => {
+  // Remove whitespaces from the beginning and end of the email address and lower case it
+  let email = emailString.trim().toLowerCase();
+
+  // Remove everything between “+” and “@”
+  email = email.replace(/\+[^@]+/g, '');
+
+  // Remove any periods that come before “@”
+  email = email.replace(/\./g, (match, offset) => {
+    if (offset < email.indexOf('@')) {
+      return '';
+    }
+    return match;
+  });
+
+  /* Make sure 
+  1. email address contains “@” sign 
+  2. there is a period after “@” 
+  3. it doesn't start or end with a period
+  */
+  if (
+    !email.includes('@') ||
+    !/\.\w+$/.test(email) ||
+    email.startsWith('.') ||
+    email.endsWith('.')
+  ) {
+    return undefined;
+  }
+
+  // Remove any spaces
+  email = email.replace(/\s/g, '');
+
+  // Remove any accents (ex: à)
+  email = email.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
+
+  // Calculate and return the SHA256 hash of the final email address
+  return sha256(email).toString();
+};
+
+/**
+ * Format and Hash PhoneNumber
+ * @param {*} phoneNumber
+ * @returnsformatted and hashed phone number
+ */
+function formatAndHashPhoneNumber(phoneNumber) {
+  // Remove any non-digit characters
+  const cleanedPhoneNumber = phoneNumber.replace(/\D/g, '');
+
+  // Check if the phone number starts with a '+' sign
+  const formattedPhoneNumber = cleanedPhoneNumber.startsWith('+')
+    ? cleanedPhoneNumber
+    : `+${cleanedPhoneNumber}`;
+
+  // Calculate and return the SHA256 hash of the formatted phone number
+  return sha256(formattedPhoneNumber).toString();
+}
+
+/**
+ * Constructing Pid Payload
+ * @param {*} context
+ */
+const constructPidPayload = message => {
+  const email = get(message, 'context.traits.email') || get(message, 'traits.email');
+  const phone = get(message, 'context.traits.phone') || get(message, 'traits.phone');
+  const pid = {};
+  if (isDefinedAndNotNull(email)) {
+    pid.em = formatAndHashEmailAddress(email);
+  }
+  if (isDefinedAndNotNull(phone)) {
+    pid.ph = formatAndHashPhoneNumber(phone);
+  }
+  /* Bing ads says if anyone one of the properties is not available 
```

Comment:
@anantjain45823 can you link public documents here

Reply:
done